diff --git a/include/memory.hpp b/include/memory.hpp
--- a/include/memory.hpp
+++ b/include/memory.hpp
@@ -54,8 +54,8 @@ namespace Kadath {

 class coef_mem {
   private:
-    static std::array<std::unique_ptr<double>,2> mem_ptrs;
-    static std::array<size_t,2> lengths;
+    static thread_local std::array<std::unique_ptr<double>,2> mem_ptrs;
+    static thread_local std::array<size_t,2> lengths;

   public:
     static double* get_mem(size_t const num, size_t const len) {
@@ -81,7 +81,7 @@ class MemoryMapper {

     using mem_map_t = std::vector<ptr_vec_t>;
     using mem_sizes_t = std::vector<size_t>;
-    static mem_sizes_t mem_sizes;
+    static thread_local mem_sizes_t mem_sizes;

     #else

@@ -92,8 +92,8 @@ class MemoryMapper {
     using ptr_t = std::unique_ptr<void,FreeDeleter>;
     using ptr_list_t = std::vector<ptr_t>;

-    static mem_map_t mem_map;
-    static ptr_list_t ptr_list;
+    static thread_local mem_map_t mem_map;
+    static thread_local ptr_list_t ptr_list;

   public:

diff --git a/src/Coef/coef_1d.cpp b/src/Coef/coef_1d.cpp
--- a/src/Coef/coef_1d.cpp
+++ b/src/Coef/coef_1d.cpp
@@ -44,8 +44,8 @@ fftw_precomp_t& coef_1d_fftw(int n) {
 }


-std::array<std::unique_ptr<double>,2> coef_mem::mem_ptrs;
-std::array<size_t,2> coef_mem::lengths {};
+thread_local std::array<std::unique_ptr<double>,2> coef_mem::mem_ptrs;
+thread_local std::array<size_t,2> coef_mem::lengths {};

 void coef_i_1d (int, Array<double>&) ;

diff --git a/src/Utilities/memory.cpp b/src/Utilities/memory.cpp
--- a/src/Utilities/memory.cpp
+++ b/src/Utilities/memory.cpp
@@ -21,11 +21,11 @@

 namespace Kadath {

-MemoryMapper::mem_map_t MemoryMapper::mem_map;
-MemoryMapper::ptr_list_t MemoryMapper::ptr_list;
+thread_local MemoryMapper::mem_map_t MemoryMapper::mem_map;
+thread_local MemoryMapper::ptr_list_t MemoryMapper::ptr_list;

 #ifdef KADATH_VECTORMAP
-MemoryMapper::mem_sizes_t MemoryMapper::mem_sizes;
+thread_local MemoryMapper::mem_sizes_t MemoryMapper::mem_sizes;
 #endif

-}
\ No newline at end of file
+}

diff --git a/include/coord_fields.hpp b/include/coord_fields.hpp
--- a/include/coord_fields.hpp
+++ b/include/coord_fields.hpp
@@ -389,11 +389,11 @@ void update_fields (CoordFields<space_t> const & cf_generator,
                    System_of_eqs* syst) {
 
   if(coord_vectors[GLOBAL_ROT]) 
-    *coord_vectors[GLOBAL_ROT] = cf_generator.template rot_z(xo);
-	if(coord_vectors[BCO1_ROT])   
-    *coord_vectors[BCO1_ROT]   = cf_generator.template rot_z(xc1) ;
-	if(coord_vectors[BCO2_ROT])   
-    *coord_vectors[BCO2_ROT]   = cf_generator.template rot_z(xc2);
+    *coord_vectors[GLOBAL_ROT] = cf_generator.rot_z(xo);
+	if(coord_vectors[BCO1_ROT])
+    *coord_vectors[BCO1_ROT]   = cf_generator.rot_z(xc1) ;
+	if(coord_vectors[BCO2_ROT])
+    *coord_vectors[BCO2_ROT]   = cf_generator.rot_z(xc2);
 	if(coord_vectors[EX])
     *coord_vectors[EX]         = cf_generator.template e_cart<COV>(1);
 	if(coord_vectors[EY])         
